src/NSubstitute/Core/Extensions.cs

```diff
@@ -24,17 +24,44 @@ public static bool IsCompatibleWith(this object? instance, Type type)
         }
 
         var instanceType = instance.GetType();
+        var compatibleInstanceTypes = GetCompatibleTypes(instanceType);
 
-        if (instanceType.IsGenericType && type.IsGenericType
-                && instanceType.GetGenericTypeDefinition() == type.GetGenericTypeDefinition())
+        foreach (var aCompatibleInstanceType in compatibleInstanceTypes)
         {
-            // both are the same generic type. If their GenericTypeArguments match then they are equivalent 
-            return CallSpecification.TypesAreAllEquivalent(instanceType.GenericTypeArguments, type.GenericTypeArguments);
+            if (aCompatibleInstanceType.IsGenericType &&
+                type.IsGenericType &&
+                aCompatibleInstanceType.GetGenericTypeDefinition() == type.GetGenericTypeDefinition())
```

Comment:
**question:** does it make any performance difference if we pull `type.GetGenerateTypeDefinition` out of the loop?

Maybe:

```csharp
var genericType = type.IsGenericType ? type.GetGenericTypeDefinition() : null;

foreach (...) {
    if (aCompatibleInstanceType.IsGenericType &&  aCompatibleInstanceType.GetGenericTypeDefinition() == genericType) ...
}
```